My take on a Focus refactor

algernon opened the pull request:
In a similar - but perhaps bolder - spirit as #1230, this is a refactor of some parts of `Focus`, namely the parts around `onFocusEvent`. It introduces a number of new helper functions, and replaces a number of patterns with something that I believe to be better.

The refactor is likely best explained by a few examples that show the old and new ways. The old ways still work, mind you, compatibility is not broken, but the new patterns are more efficient.

## `Focus.handleHelp` => `.inputMatchesHelp()` + `.printHelp()`

Old:

```c++
  if (::Focus.handleHelp(command, PSTR("autoshift.enabled\r\n"
                                       "autoshift.timeout\r\n"
                                       "autoshift.categories")))
    return EventHandlerResult::OK;
```

New:

```c++
  const char *cmd_enabled    = PSTR("autoshift.enabled");
  const char *cmd_timeout    = PSTR("autoshift.timeout");
  const char *cmd_categories = PSTR("autoshift.categories");

  if (::Focus.inputMatchesHelp(command))
    return ::Focus.printHelp(cmd_enabled, cmd_timeout, cmd_categories);
```

## Subcommands

Old:

```c++
  if (strncmp_P(command, PSTR("autoshift."), 10) != 0)
    return EventHandlerResult::OK;
  if (strcmp_P(command + 10, PSTR("enabled")) == 0)
    subCommand = ENABLED;
  else if (strcmp_P(command + 10, PSTR("timeout")) == 0)
    subCommand = TIMEOUT;
  else if (strcmp_P(command + 10, PSTR("categories")) == 0)
    subCommand = CATEGORIES;
  else
    return EventHandlerResult::OK;
```

New:

```c++
  if (::Focus.inputMatchesCommand(command, cmd_enabled))
    subCommand = ENABLED;
  else if (::Focus.inputMatchesCommand(command, cmd_timeout))
    subCommand = TIMEOUT;
  else if (::Focus.inputMatchesCommand(command, cmd_categories))
    subCommand = CATEGORIES;
  else
    return EventHandlerResult::OK;
```

There are multiple wins here.

First, we no longer need to hand-write the help message. We can reuse the same strings used for the help for detecting the commands we handle. This reduces duplicated strings considerably.

Second, we no longer need to check for a prefix first (an error prone operation previously, having to maintain the prefix length for the `strncmp_P` call), we can just check the full command against the strings we used for `help`. This makes the flow of code shorter, and cleaner.

Third, since we no longer need the prefix check, that's both code dropped, and smaller binary size.

## `inputMatchesCommand`

This helper is also used above, but it doesn't depend on subcommands, can be used anytime we want to check if a particular input matches a command.

# Size matters!

Comparing the sketches of `Chrysalis-Firmware-Bundle`:

- ErgoDox is down 72 bytes
- Keyboardio Atreus is down by 80 bytes
- Keyboardio Model01 is down by 34 bytes
- Keyboardio Model100 is up by 160 bytes (but the Model100 is unoptimized, and has a lot of space, so doesn't matter much at this point)
- Splitography is down by 72 bytes
- Original Atreus is down by 72 bytes

Not a lot of bytes saved, but considering that the code got cleaner, and easier, and our size went down, rather than up, I'll call that a big win anyway.

# TODO

- [ ] Review the naming. Not entirely happy with `.inputMatchesCommand(command, cmd_something)`.
- [ ] Clean up the history
- [ ] Update the FocusSerial README to show the new helpers
- [ ] Update `UPGRADING.md` too

obra commented:
>  Review the naming. Not entirely happy with .inputMatchesCommand(command, cmd_something).

Indeed, but `.inputMatchesCommand(input, cmd_something)` feels pretty reasonable to me.

This seems like a very reasonable improvement.

algernon commented:
> > Review the naming. Not entirely happy with .inputMatchesCommand(command, cmd_something).
> 
> Indeed, but `.inputMatchesCommand(input, cmd_something)` feels pretty reasonable to me.
> 
> This seems like a very reasonable improvement.

Hm. We'll need to have a terminology pass on Focus then, because `command` here is `onFocusEvent(const char *command)`. We'd need to rename it there. Could work, could work. I'll do that and see how it feels.

algernon commented:
- Switched from `onFocusEvent(command)` to `onFocusEvent(input)` - just renamed the argument.
- Added the new helpers to the `FocusSerial` README
- Updated `UPGRADING.md` to show the new APIs. I did not document how to transition from `handleHelp()`, not sure what would be the best place for that.

I'll clean up the commit history soon.

tlyu commented on plugins/Kaleidoscope-FocusSerial/src/kaleidoscope/plugin/FocusSerial.h:
```diff
@@ -26,6 +26,15 @@
+#include "kaleidoscope_internal/deprecations.h"  // for DEPRECATED
+
+#define _DEPRECATED_MESSAGE_FOCUS_HANDLEHELP                      \
+  "The `Focus.handeHelp()` method is deprecated. Please use\n"    \
```
typo: should probably be `Focus.handleHelp()`

algernon replied:
Fixed, thanks!

tlyu reviewed (commented):
I have read the changes but not compiled or tested them.

Overall, I think it's a substantial improvement. The error-prone hardcoded offset for command prefixes is gone, as is duplication of help and command strings.

On the downside, there is increased duplication of command prefix strings. It might be possible to mitigate this by adding logic in `FocusSerial` to handle command prefixes, so that each plugin doesn't need to do so separately. If there's a net code size savings, the current changes might be good enough for now.

I do have some ideas for further improvements, but they might impose additional constraints on the command structure available to plugin authors. Would you like me to write them up in this PR, or in a separate issue?

algernon commented:
> On the downside, there is increased duplication of command prefix strings. It might be possible to mitigate this by adding logic in `FocusSerial` to handle command prefixes, so that each plugin doesn't need to do so separately. If there's a net code size savings, the current changes might be good enough for now.

I've done some local tests, to see if lifting out the prefixes would help, and on AVR, that ended up compiling to a larger size, because the extra code required for the prefix handling was more than what it saved. It's very likely that it could be done in a better way than I tried, and save some bytes, but I do not think it's worth it. This PR already results in 30-80 bytes saved on AVR boards (using the sketches from Chrysalis-Firmware-Bundle), *and* easier to read code on top.

Even if we could save a couple of more bytes, I don't see how that'd improve code readability. I also think that AVR is not likely to be used in future builds much, so code size doesn't matter *that* much anymore, so we can make little sacrifices that may result in larger size, but let us write easier to follow code.

> I do have some ideas for further improvements, but they might impose additional constraints on the command structure available to plugin authors. Would you like me to write them up in this PR, or in a separate issue?

I'd love to hear your ideas. I think a separate issue would be better, so it's not tied to this PR, but rather is a followup.

tlyu commented on plugins/Kaleidoscope-FocusSerial/src/kaleidoscope/plugin/FocusSerial.h:
```diff
@@ -37,8 +46,22 @@ class FocusSerial : public kaleidoscope::Plugin {
+  }
+  template<typename... Vars>
+  EventHandlerResult printHelp(const char *h1, Vars... vars) {
+    Runtime.serialPort().println((const __FlashStringHelper *)h1);
```
Do you maybe want to add a `delayAfterPrint()` here?

algernon replied:
I do. Fixed, thanks!

algernon commented:
History is now cleaned up, so this is ready for final review.

obra reviewed (approved):
I'm good with this. I wish we could have tests for it, but that's not a blocker
